### PR TITLE
[Merged by Bors] - feat: growth of a finset in the quotient and intersection with a subgroup

### PR DESCRIPTION
**Pull request by YaelDillies**
From GrowthInGroups (LeanCamCombi)

---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/Combinatorics/Additive/QuotientInterGrowth.lean [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+
+open Finset Function
+open scoped Pointwise
+
+namespace Finset
+variable {G : Type*} [Group G] [DecidableEq G] {H : Subgroup G} [DecidablePred (· ∈ H)] [H.Normal]
+  {A : Finset G} {m n : ℕ}
+
+@[to_additive]
+lemma card_pow_quotient_mul_pow_inter_subgroup_le :
```
Comment (b-mehta) [inline Mathlib/Combinatorics/Additive/QuotientInterGrowth.lean:25]: I agree this is morally an `inter`, but if we're using this naming convention then I'm not sure `inter` is appropriate here...

File: Mathlib/Combinatorics/Additive/QuotientInterGrowth.lean [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+Authors: Yaël Dillies
+-/
+import Mathlib.Algebra.Group.Pointwise.Finset.Basic
+import Mathlib.Combinatorics.Enumerative.DoubleCounting
+import Mathlib.GroupTheory.QuotientGroup.Defs
+
+/-!
+# Growth in the quotient and intersection with a subgroup
+
+For a group `G` and a subgroup `H ≤ G`, this file upper and lower bounds the growth of a finset by
```
Comment (b-mehta) [inline Mathlib/Combinatorics/Additive/QuotientInterGrowth.lean:13]: It looks like you only lower bound the growth currently?

File: Mathlib/Combinatorics/Additive/QuotientInterGrowth.lean [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+
+@[to_additive]
+lemma le_card_quotient_mul_sq_inter_subgroup (hAsymm : A⁻¹ = A) :
+    #A ≤ #(A.image <| QuotientGroup.mk' H) * #{x ∈ A ^ 2 | x ∈ H} := by
+  classical
+  set π := QuotientGroup.mk' H
+  calc
+    #A = #A * 1 := by rw [mul_one]
+    _ ≤ #(A.image π) * #{x ∈ A ^ 2 | x ∈ H} :=
+      card_mul_le_card_mul (π · = ·) (by simp [Finset.Nonempty]; aesop) ?_
```
Comment (b-mehta) [inline Mathlib/Combinatorics/Additive/QuotientInterGrowth.lean:65]:
```suggestion
      card_mul_le_card_mul (π · = ·) (by aesop (add simp Finset.Nonempty)) ?_
```
maybe?

File: Mathlib/Combinatorics/Additive/QuotientInterGrowth.lean [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+  simp only [mem_image, forall_exists_index, and_imp, forall_apply_eq_imp_iff₂]
+  rintro a ha
+  calc
+    #(bipartiteBelow (π · = ·) A (π a))
```
Comment (b-mehta) [inline Mathlib/Combinatorics/Additive/QuotientInterGrowth.lean:69]: In this specific case isn't it easier to use one of the lemmas about fibres?

**Review by b-mehta — APPROVED** [review]
Nice!

bors d+

File: Mathlib/Combinatorics/Additive/QuotientInterGrowth.lean (hunk 3 above)
Comment (YaelDillies) [reply Mathlib/Combinatorics/Additive/QuotientInterGrowth.lean:65]: Nope, times out

File: Mathlib/Combinatorics/Additive/QuotientInterGrowth.lean (hunk 1 above)
Comment (YaelDillies) [reply Mathlib/Combinatorics/Additive/QuotientInterGrowth.lean:25]: I'm following the naming conventions from the `GroupTheory.QuotientGroup` folder, but I agree they aren't great... I don't have a better idea here though

File: Mathlib/Combinatorics/Additive/QuotientInterGrowth.lean (hunk 4 above)
Comment (YaelDillies) [reply Mathlib/Combinatorics/Additive/QuotientInterGrowth.lean:69]: don't really see how, no

File: Mathlib/Combinatorics/Additive/QuotientInterGrowth.lean (hunk 2 above)
Comment (YaelDillies) [reply Mathlib/Combinatorics/Additive/QuotientInterGrowth.lean:13]: `card_pow_quotient_mul_pow_inter_subgroup_le` is the lower bound, and `le_card_quotient_mul_sq_inter_subgroup` is the upper bound (just set `A := A ^ n`)

File: Mathlib/Geometry/Group/Growth/QuotientInter.lean [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+      card_mul_le_card_mul (π · = ·) (by simp [Finset.Nonempty]; aesop) ?_
+  simp only [mem_image, forall_exists_index, and_imp, forall_apply_eq_imp_iff₂]
+  rintro a ha
+  calc
+    #(bipartiteBelow (π · = ·) A (π a))
+    _ ≤ #((bipartiteBelow (π · = ·) A (π a))⁻¹ * (bipartiteBelow (π · = ·) A (π a))) :=
+      card_le_card_mul_left _ ⟨a⁻¹, by simpa⟩
+    _ ≤ #{x ∈ A⁻¹ * A | x ∈ H} := by
+      gcongr
+      simp only [mul_subset_iff, mem_inv', mem_bipartiteBelow, map_inv, Finset.mem_filter, and_imp]
```
Comment (b-mehta) [inline Mathlib/Geometry/Group/Growth/QuotientInter.lean:74]:
```suggestion
      simp only [mul_subset_iff, mem_inv', mem_bipartiteBelow, map_inv, mem_filter, and_imp]
```

File: Mathlib/Combinatorics/Additive/QuotientInterGrowth.lean (hunk 4 above)
Comment (b-mehta) [reply Mathlib/Combinatorics/Additive/QuotientInterGrowth.lean]: This is a golf, and avoids the simp and aesop earlier. Plus it means you can also remove the `mem_bipartiteBelow` later, and write the first rhs in the second calc more clearly: `#({b ∈ A | π b = π a}⁻¹ * {b ∈ A | π b = π a})`

Comment (YaelDillies) [reply Mathlib/Combinatorics/Additive/QuotientInterGrowth.lean]: Oh I see! I had no idea about `card_eq_sum_card_image`

**Comment by YaelDillies** [comment]
bors merge

**Comment by YaelDillies** [comment]
bors merge

**Pull request closed without merging**